go/cmd/dolt/commands/blame.go

```diff
@@ -0,0 +1,450 @@
+		return nil, err
+	}
+
+	blameGraph, err := blameGraphFromRows(ctx, rows)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, c := range commits {
+		// get the first parent
+		parent, err := dEnv.DoltDB.ResolveParent(ctx, c, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		// for each unblamed node, see if it changed between `c` and `parent`.
+		// if so, mark it blamed with `c` as the blame origin
+		for _, rowPK := range blameGraph.UnblamedNodeKeys() {
+			changed, err := rowChanged(ctx, parent, c, tableName, rowPK)
+			if err != nil {
+				return nil, err
+			}
+
+			if changed {
+				blameGraph.BlameNode(rowPK, c)
+			}
+		}
+
+		// if all nodes have blame, stop iterating commits
+		if blameGraph.AllNodesBlamed() {
+			break
+		}
+	}
+	return blameGraph, nil
+}
+
+// rowsFromCommit returns the row data of the table with the given name at the given commit
+func rowsFromCommit(ctx context.Context, commit *doltdb.Commit, tableName string) (types.Map, error) {
+	root, err := commit.GetRootValue()
+	if err != nil {
+		return types.EmptyMap, err
+	}
+
+	table, ok, err := root.GetTable(ctx, tableName)
+	if err != nil {
+		return types.EmptyMap, err
+	}
+	if !ok {
+		return types.EmptyMap, fmt.Errorf("no table named %s found", tableName)
+	}
+
+	rowData, err := table.GetRowData(ctx)
+	if err != nil {
+		return types.EmptyMap, err
+	}
+
+	return rowData, nil
+}
+
+func getCommitHashes(old, new *doltdb.Commit) (string, string, error) {
+	oldHash, err := old.HashOf()
+	if err != nil {
+		return "", "", fmt.Errorf("error getting hash of old commit: %v", err)
+	}
+	newHash, err := new.HashOf()
+	if err != nil {
+		return "", "", fmt.Errorf("error getting hash of new commit: %v", err)
+	}
+	return oldHash.String(), newHash.String(), nil
+}
+
+// maybeTableFromCommit takes a commit and a table name and returns a (possibly nil) pointer to a table
+func maybeTableFromCommit(ctx context.Context, c *doltdb.Commit, tableName string) (*doltdb.Table, error) {
+	root, err := c.GetRootValue()
+	if err != nil {
+		return nil, fmt.Errorf("error getting root value of commit: %v", err)
+	}
+	table, _, err := root.GetTable(ctx, tableName)
+	if err != nil {
+		return nil, fmt.Errorf("error getting table %s from root value: %v", tableName, err)
+	}
+	return table, nil
+}
+
+// maybeRowFromTable takes a table and a primary key and returns a (possibly nil) pointer to a row
+func maybeRowFromTable(ctx context.Context, t *doltdb.Table, rowPK types.Value) (*row.Row, error) {
+	schema, err := t.GetSchema(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting schema from table: %v", err)
+	}
+
+	row, ok, err := t.GetRow(ctx, rowPK.(types.Tuple), schema)
+	if err != nil {
+		return nil, fmt.Errorf("error getting row from table: %v", err)
+	}
+	if !ok {
+		return nil, nil
+	}
+
+	return &row, err
+}
+
+func schemaFromCommit(ctx context.Context, c *doltdb.Commit, tableName string) (schema.Schema, error) {
+	t, err := maybeTableFromCommit(ctx, c, tableName)
```

Comment:
Would be great to package these (`schema` and `t`) up with the `commits` slice we're ranging up in `blameGraphFromCommit` since we're doing this same work for every row * every commit.

Reply:
Yeah I know there's a decent bit of extra work being done in this function. If things are looking good overall I'm happy to start zeroing in on optimizations and the like.

Reply:
Okay, a bunch of inputs now get precomputed and packaged: https://github.com/liquidata-inc/dolt/compare/0665eb6fa9aff9442d12f20c33532893ee7e16f7...2721ea4

Reply:
Cool!